classes/patreon_protect.php

```diff
@@ -0,0 +1,812 @@
+			
+			$y1_coord = $half_height + ceil( $font_size / 2 ) + 25;
+			
+			$x2_coord = $width-( $width - $rectangle_width ) / 2;
+			$y2_coord = $y1_coord + $rectangle_height;
+			
+			// Nudge up if it text is hidden
+
+			if ( isset( $hide_text ) ) {
+				$y1_coord = $half_height + 10;
+				$y2_coord = $y1_coord + $rectangle_height;
+			}
+			
+			$thickness = apply_filters( 'ptrn/locked_image_interface_button_border', 3, $patreon_level, $attachment_id, $image, $width, $height);
+
+			imagesetthickness( $image, $thickness );
+
+			imagerectangle( $image, $x1_coord, $y1_coord, $x2_coord, $y2_coord, $white );
+
+			$button_font_size = apply_filters( 'ptrn/locked_image_interface_unlock_button_font_size', 9, $patreon_level, $attachment_id, $image, $width, $height );
+
+			$dimensions  = imagettfbbox( $button_font_size, 0, $font, $unlock_button_text );
+
+			$text_width  = $dimensions[2] - $dimensions[0];
+			$text_height = $dimensions[7] - $dimensions[1];
+
+			$x_coord = ( $width - $text_width ) / 2;
+
+			$y_coord = $y1_coord + ( ( $rectangle_height / 2 ) - ( $text_height / 2 ) );
+
+			imagettftext( $image, $button_font_size, 0, $x_coord, $y_coord, $white, $font, $unlock_button_text );
+		}
+		
+		header( 'Content-Type: '.$force_mime_type );
+
+		if ( $force_mime_type=='image/png' ) {
+			imagepng( $image,PATREON_PLUGIN_LOCKED_IMAGE_CACHE_DIR . '/' . $cached_filename );
+		}
+		if ( $force_mime_type=='image/gif' ) {
+			imagegif( $image,PATREON_PLUGIN_LOCKED_IMAGE_CACHE_DIR . '/' . $cached_filename );
+		}
+		if ( $force_mime_type=='image/jpeg' ) {
+			imagejpeg( $image,PATREON_PLUGIN_LOCKED_IMAGE_CACHE_DIR . '/' . $cached_filename );
+		}
+		if ( $force_mime_type=='image/bmp' ) {
+			imagebmp( $image, PATREON_PLUGIN_LOCKED_IMAGE_CACHE_DIR . '/' . $cached_filename );
+		}
+		
+		// Readfile below for lower memory usage. Can be changed to echo file_get_contents for small images in future
+		echo file_get_contents( PATREON_PLUGIN_LOCKED_IMAGE_CACHE_DIR . '/' . $cached_filename );
+			
+		imagedestroy( $image );
+	}
+	public static function addPatreonRewriteRules() {
+			
+		// File locking not enabled. Return
+		if ( get_option( 'patreon-enable-file-locking', false ) == false ) {
+			return;
+		}
+		
+		$htaccess = file_get_contents( ABSPATH . '.htaccess' );
+		
+		$upload_locations = wp_upload_dir();
+
+		// We want the base upload location so we can account for any changes to date based subfolders in case there are
+
+		$upload_dir = substr( wp_make_link_relative( $upload_locations['baseurl'] ) , 1 );
+
+		$append = PHP_EOL . "# BEGIN Patreon WordPress Image Protection
+RewriteEngine On
+RewriteBase /		
+RewriteCond %{REQUEST_FILENAME} (\.png|\.jpg|\.gif|\.jpeg|\.bmp)
+RewriteCond %{HTTP_REFERER} !^wp-admin [NC]
+RewriteRule ^" . $upload_dir . "/(.*)$ index.php?patreon_action=serve_patron_only_image&patron_only_image=$1 [QSA,L]
+# END Patreon WordPress".PHP_EOL;
+		
+    	file_put_contents( ABSPATH .'.htaccess', $htaccess . $append );

```

Comment:
indent?

Reply:
Since its text wrapped in EOL, if we indent the text added to htaccess is going to indent itself inside htaccess too. Which will look weird since it is a block in itself and not contained in any other block. We should leave it as it is.

Reply:
Sounds good 👍  Thanks Oz!